new-integration-tests/src/test/java/oracle/weblogic/kubernetes/assertions/impl/Pod.java

```diff
@@ -0,0 +1,200 @@
+      return false;
+    } else {
+      logger.info("WebLogic pods found in namespace {0}", namespace);
+      for (V1Pod item : listPods.getItems()) {
+        logger.info(item.getMetadata().getName());
+        podNames.add(item.getMetadata().getName());
+      }
+    }
+
+    // check the pods termination status in concurrent thread
+    ExecutorService executorService = Executors.newFixedThreadPool(podNames.size());
+    ArrayList<Future<Boolean>> threads = new ArrayList<Future<Boolean>>();
+    for (var podName : podNames) {
+      // check for pod termination status and return true if pod is terminating
+      threads.add(executorService.submit(() -> {
+        retry
+            .conditionEvaluationListener(
+                condition -> logger.info("Waiting for pod {0} in namespace {1} to terminate"
+                    + "(elapsed time {2}ms, remaining time {3}ms)",
+                    podName,
+                    namespace,
+                    condition.getElapsedTimeInMS(),
+                    condition.getRemainingTimeInMS()))
+            .until(onlyGivenPodTerminating(podName, domainUid, namespace));
+        return true;
+      }));
+      // wait for the callable to finish running and check if all pods were terminating
+      for (var future : threads) {
+        if (!future.get(10, MINUTES)) {
+          return false;
+        }
+      }
+    }
+    executorService.shutdownNow();
+
+    // wait for pods to become ready
+    for (var podName : podNames) {
+      logger.info("Wait for pod {0} to be ready in namespace {1}", podName, namespace);
+      retry
+          .conditionEvaluationListener(
+              condition -> logger.info("Waiting for pod {0} to be ready in namespace {1} "
+                  + "(elapsed time {2}ms, remaining time {3}ms)",
+                  podName,
+                  namespace,
+                  condition.getElapsedTimeInMS(),
+                  condition.getRemainingTimeInMS()))
+          .until(assertDoesNotThrow(() -> podReady(namespace, domainUid, podName),
+              String.format(
+                  "pod %s is not ready in namespace %s", podName, namespace)));
+    }
+
+    return true;
+  }
+
+  /**
+   * Return true if the given pod is the only one terminating.
+   *
+   * @param podName name of pod to check for termination status
+   * @param domainUid UID of the WebLogic domain
+   * @param namespace name of the namespace in which the pod is running
+   * @return true if given pod is terminating otherwise false
+   * @throws Exception when more than one pod is terminating or cluster query fails
+   */
+  private static Callable<Boolean> onlyGivenPodTerminating(String podName, String domainUid, String namespace)
+      throws Exception {
+    return () -> {
+      String labelSelectors = String.format("weblogic.serverName", domainUid);
+      V1PodList listPods = Kubernetes.listPods(namespace, labelSelectors);
+      if (listPods.getItems().isEmpty()) {
```

Comment:
I could be wrong, but I suspect there may be multiple issues with this assertion, some of it functional, some of it due to incompleteness.  All of the following should be double checked with Ryan, Lenny, and Anthony:

- Seems to assume a single cluster and no non-clustered servers. Consequently, this test seems to assume that only one server can go down a time.  This is incorrect when there are multiple clusters, unmanaged servers, a replica count change, or a deleted cluster.

- Seems to dynamically obtain a list of servers when checking if only one server is 'down'.  IMO, it'd be better to get the list of servers at the beginning of a test before any changes are made to initiate the roll (otherwise, it won't catch if two servers in the cluster 'disappear' at the same time).  This 'original list' should be then compared to what's expected by the domain resource in terms of replicas, restartVersions, and image in order to determine the expected behavior for the roll.

- Note that this may get a little tricky when 'replicas' changes at the same time as everything else.  When replicas increases, the number of 'not ready' can be higher than 1 since the replicas are allowed to start in parallel.   When replicas decreases I assume (but am not sure) the roll is expected to defer terminating the servers until _after_ it's completed the roll of the previous servers, and then only terminate them one at a time.   Similar issues may be seen when a cluster is deleted (the roll doesn't apply in that case I assume) or a new cluster is added (multiple servers can start at once).

- Seems to use 'Terminating' as a criteria when checking if a single server is down. For any particular domain it should check the admin server is ready - if not, then the admin server should be the only cycling server.  Then it should check that all other servers are 'ready' except one for their particular cluster as compared to the list of pods obtained at the beginning of the test (except in the case of varying replicas).
- Does not seem to verify that admin server always is 'ahead' of other servers (first to roll and reach target state, and only one that should be down until this occurs)?
- Does not verify that rolls are expected to occur 'in order' within a given cluster.
- At the conclusion of the test, does not check if all servers have (a) reached their target domain/cluster/server restartVersion and (b) reached their target image.  (This should be done in addition to checking for  'ready' state.)

Reply:
Removed getting pods dynamically and instead the pod names are passed to the assertion method as a list. The assertion method will simply go through the list and verify its restarted in the sequence it is added in the list, at the same verifies no other pods are restarted at the same time. verifying the restart version after rolling should be done in the test IMO.

Reply:
> verifying the restart version after rolling should be done in the test IMO.

Reply:
Why? A "proper" restart version and image are required for every single roll.  They're essential for corrects behavior.  So why not centrally ensure they match the domain resource's values for every roll?

Reply:
The requirement is to assert pods replacement doesn't exceed maxUnavailable at a time. We can create another assertion to ensure the integrity of the restartVersion and image.

Reply:
the 'spirit' of the requirement IMO, is to check that roll is working as designed from a customer perspective, this includes restart-version and image